video/out/vo_sixel.c

```diff
@@ -0,0 +1,429 @@
+    int diffuse;
+    int width;
+    int height;
+    int reqcolors;
+    int fixedpal;
+    int threshold;
+    int top;
+    int left;
+};
+
+#define OPT_BASE_STRUCT struct vo_sixel_opts
+static const struct m_sub_options vo_sixel_conf = {
+    .opts = (const m_option_t[]) {
+        {"vo-sixel-diffusion", OPT_CHOICE(diffuse,
+            {"auto", DIFFUSE_AUTO},
+            {"none", DIFFUSE_NONE},
+            {"atkinson", DIFFUSE_ATKINSON},
+            {"fs", DIFFUSE_FS},
+            {"jajuni", DIFFUSE_JAJUNI},
+            {"stucki", DIFFUSE_STUCKI},
+            {"burkes", DIFFUSE_BURKES},
+            {"arithmetic", DIFFUSE_A_DITHER},
+            {"xor", DIFFUSE_X_DITHER})},
+        {"vo-sixel-width", OPT_INT(width)},
+        {"vo-sixel-height", OPT_INT(height)},
+        {"vo-sixel-reqcolors", OPT_INT(reqcolors)},
+        {"vo-sixel-fixedpalette", OPT_INT(fixedpal)},
+        {"vo-sixel-color-threshold", OPT_INT(threshold)},
+        {"vo-sixel-offset-top", OPT_INT(top)},
+        {"vo-sixel-offset-left", OPT_INT(left)},
+        {0}
+    },
+    .defaults = &(const struct vo_sixel_opts) {
+        .diffuse = DIFFUSE_ATKINSON,
+        .width = 320,
+        .height = 240,
+        .reqcolors = 256,
+        .fixedpal = 0,
+        .threshold = 0,
+        .top = 1,
+        .left = 1,
+    },
+    .size = sizeof(struct vo_sixel_opts),
+};
+
+struct priv {
+    struct vo_sixel_opts *opts;
+
+    sixel_output_t *output;
+    sixel_dither_t *dither;
+    sixel_dither_t *testdither;
+    uint8_t        *buffer;
+
+    int image_height;
+    int image_width;
+    int image_format;
+
+    struct mp_image *frame;
+    struct mp_sws_context *sws;
+};
+
+static const unsigned int depth = 3;
+
+static void save_image_offset_location(struct vo* vo)
+{
+    struct priv* priv = vo->priv;
+    int top = priv->opts->top;
+    int left = priv->opts->left;
+    int terminal_width = 0;
+    int terminal_height = 0;
+
+    terminal_get_size(&terminal_width, &terminal_height);
+
+    // Make sure that the user specified top offset
+    // lies in the range 1 to TERMINAL_HEIGHT
+    // Otherwise default to the topmost row
+    if (top <= 0 || top > terminal_height)
+        top = 1;
+
+    // Make sure that the user specified left offset
+    // lies in the range 1 to TERMINAL_WIDTH
+    // Otherwise default to the leftmost column
+    if (left <= 0 || top > terminal_width)
+        left = 1;
+
+    /*We want to save the offset row and column
+    at which the sixel output will be displayed.
+    Controlled by the --vo-sixel-offset- arguments.*/
+    printf(ESC_SAVE_XY_CURSOR_LOCATION, top, left);
+}
+
+static int detect_scene_change(struct vo* vo)
+{
+    struct priv* priv = vo->priv;
+    int score;
+    int i;
+    unsigned int r = 0;
+    unsigned int g = 0;
+    unsigned int b = 0;
+    static unsigned int average_r = 0;
+    static unsigned int average_g = 0;
+    static unsigned int average_b = 0;
+    static int previous_histgram_colors = 0;
```

Comment:
static variables are a big no-no

Reply:
I'll move them to `struct priv`, shouldn't be a big change.